### PR TITLE
bpo-41229: Update docs for explicit aclose()-required cases and add contextlib.aclosing() method

### PR DESCRIPTION
**Pull request by achimnol**
This is a PR to:

 * Add `contextlib.aclosing` which ia analogous to `contextlib.closing` but for async-generators with an explicit test case for [bpo-41229](https://bugs.python.org/issue41229)
 * Update the docs to describe when we need explicit `aclose()` invocation.

which are motivated by the following issues, articles, and examples:

 * [bpo-41229](https://bugs.python.org/issue41229)
 * https://github.com/njsmith/async_generator
 * https://vorpus.org/blog/some-thoughts-on-asynchronous-api-design-in-a-post-asyncawait-world/#cleanup-in-generators-and-async-generators
 * https://www.python.org/dev/peps/pep-0533/
 * https://github.com/achimnol/aiotools/blob/ef7bf0cea7af/src/aiotools/context.py#L152

Particuarly regarding [PEP-533](https://www.python.org/dev/peps/pep-0533/), its acceptance (`__aiterclose__()`) would make this little addition of `contextlib.aclosing()` unnecessary for most use cases, but until then this could serve as a good counterpart and analogy to `contextlib.closing()`. The same applies for `contextlib.closing` with `__iterclose__()`.
Also, still there are other use cases, e.g., when working with non-generator objects with `aclose()` methods.

https://bugs.python.org/issue41229

Automerge-Triggered-By: GH:aeros

### CODE REVIEW

**Comment by achimnol** [comment]
@njsmith I just read your comments and resolution of bpo-41229, and I also see that async-gens in the original `leak.py` were not being gc'ed because the event loop is indefinitely blocked by `input()`.
Still, I think it's worth to add this PR, so please let me know what you think about this. :)

**Comment by aeros** [comment]
Adding `DO-NOT-MERGE` for now, so it's not merged prior to approval of the `contextlib.aclosing()` feature. Without an actual memory leak occurring in async generators, this is now a more subjective enhancement to be argued for, rather than a bugfix.

(I'm not currently +1 or -1 for the feature, I have to spend some more time considering the use cases.)

**Comment by achimnol** [comment]
I think the major issue for `contextlib.aclosing` would be whether people will use the name "aclose" in their objects to async-cleanup resources.

There are currently the following approaches I see:
1. `obj.close()` – e.g., zmq.asyncio.Socket
2. `await obj.close()` – e.g., aiohttp.WebSocketResponse
3. `obj.close()` followed by `await obj.wait_closed()` – e.g., asyncio.StreamWriter
4. `await obj.aclose()` – e.g., async generator

So, from the perspective of naming, adding `contextlib.aclosing` to the stdlib might be promoting the last case for non-generator use cases as well. Supporting all 2nd, 3rd, 4th cases above in `contextlib.aclosing` would make things complicated.

Still, I believe that we need a better approach so that people don't be confused about resource cleanup upon generator termination/interruption, and currently the only way is to explicitly close the generator (for both sync/async generators).
Relying on the event loop's behavior or GC is counter-intuitive. For instance, even an experienced developer like @njsmith was surprised at his first glance on this issue and after some examination he concluded that this was "actually" an expected behavior.

I don't know how things are going on with PEP-533, but if it's going to be accepted in the future, I can agree with skipping `contextlib.aclosing`.

**Comment by 1st1** [comment]
> I think the major issue for contextlib.aclosing would be whether people will use the name "aclose" in their objects to async-cleanup resources.

We're going to start prompting `aclose()` in asyncio APIs soon.

**Review by 1st1 — APPROVED** [review]
+1 from me to add this.

**Review by corona10 — COMMENTED** [review]
@aeros 
Anything more stuff is needed before merging this PR?

**Review by aeros — CHANGES_REQUESTED** [review]
Mostly LGTM. A few nits though:

1) The docs mention `aclosing` as a function, when in `Lib/contextlib.py` it's implemented as a class. As a result, it should use the `.. class::` reST role instead of `.. function::`. Also, instead of "return an async context manager", it would be more correct to directly call it an async context manager [1]. 

2) In the docstring for `aclosing()`, I would suggest for it to be changed to something like: "Async context manager for safely finalizing an async generator using its aclose() method". I think the current version of the docstring may cause readers to miss the main point of it, and potentially make incorrect assumptions about its usage. Also, the "end of a block" part is a bit vague [2]. Instead, it makes more sense to me to just refer to it as a context manager.

3) I think this would also benefit from a "What's New" entry in 3.10 for `contextlib` to briefly mention the addition of `contextlib.aclosing()`, as it is a new public member.

Other than the  above points, I think it's good to go. Feel free to @ mention me once the changes are made, and I can proceed with merging it (unless @1st1 is 100% content with the current state, and decides to merge it as is).

---

[1] - Any class that implements `__aenter__` and `__aexit__` is considered an async context manager, or at the very least is usable as one.

[2] - I get that you're referring to the block/scope within the with statement, but I don't think that will be immediately clear to all readers; especially if they're not intimately familiar with context managers in general. By simply referring to it as an "async context manager", readers can much more easily look up how context managers work rather than trying to guess what "context" and "block" are referring to within the docstring.

**Comment by achimnol** [comment]
> Mostly LGTM. A few nits though:
> 
> 1. The docs mention `aclosing` as a function, when in `Lib/contextlib.py` it's implemented as a class. As a result, it should use the `.. class::` reST role instead of `.. function::`. Also, instead of "return an async context manager", it would be more correct to directly call it an async context manager [1].

I updated the docs to use `.. class::`. However, I see the same issue on the existing docs for `closing`, because I copy-and-pasted it to start with. Shall I update it as well?

> 2. In the docstring for `aclosing()`, I would suggest for it to be changed to something like: "Async context manager for safely finalizing an async generator using its aclose() method". I think the current version of the docstring may cause readers to miss the main point of it, and potentially make incorrect assumptions about its usage. Also, the "end of a block" part is a bit vague [2]. Instead, it makes more sense to me to just refer to it as a context manager.

Agreed and applied your suggestion, but the same question: shall I update the docstring for existing `closing()` to be consistent?

> 3. I think this would also benefit from a "What's New" entry in 3.10 for `contextlib` to briefly mention the addition of `contextlib.aclosing()`, as it is a new public member.
> 

I have already added the news fragment using blurb.
Please let me know if there are something I've missed.

> Other than the above points, I think it's good to go. Feel free to @ mention me once the changes are made, and I can proceed with merging it (unless @1st1 is 100% content with the current state, and decides to merge it as is).
> 
> [1] - Any class that implements `__aenter__` and `__aexit__` is considered an async context manager, or at the very least is usable as one.
> 
> [2] - I get that you're referring to the block/scope within the with statement, but I don't think that will be immediately clear to all readers; especially if they're not intimately familiar with context managers in general. By simply referring to it as an "async context manager", readers can much more easily look up how context managers work rather than trying to guess what "context" and "block" are referring to within the docstring.

File: Doc/library/contextlib.rst
Comment (belm0) [inline Doc/library/contextlib.rst]: should be `async def`

File: Doc/library/contextlib.rst
Comment (belm0) [inline Doc/library/contextlib.rst]: I think rather than some vague example, this is the right place to mention how it's useful for deterministic async generator cleanup, and use that as an example.

See https://github.com/python/cpython/pull/22640/files#diff-aa53aa6efed705fd0589ce0fdac405d7R171-R185

File: Doc/reference/expressions.rst
Comment (belm0) [inline Doc/reference/expressions.rst]: It could be explained more clearly-- most readers won't have in mind how this situation comes up.  And the rationale could be more detailed.

> If an asynchronous generator happens to exit early by :keyword:`break`, the caller task being cancelled, or other exception, the generator's async cleanup code will run and possibly raise exceptions or access context variables in an unexpected context-- perhaps after the lifetime of tasks it depends on or the event loop itself.  To prevent this, the caller must explicitly close the async generator by calling :meth:`~agen.aclose` method to finalize the generator and ultimately detach it from the event loop.

File: Lib/contextlib.py
Comment (belm0) [inline Lib/contextlib.py:317]: `await fetch()`

File: Misc/NEWS.d/next/Library/2020-07-19-20-10-41.bpo-41229.p8rJa2.rst
Comment (belm0) [inline Misc/NEWS.d/next/Library/2020-07-19-20-10-41.bpo-41229.p8rJa2.rst]: it would be useful to mention the primary motivation (deterministic cleanup of async generators)

**Comment by belm0** [comment]
Neglecting to mention https://bugs.python.org/issue40213 has cost me several hours of duplicated effort in https://github.com/python/cpython/pull/22640/ 😦 

Anyway I'll add comments here based on thinking I've put into it.

**Review by belm0 — CHANGES_REQUESTED** [review]

File: Doc/reference/expressions.rst
Comment (njsmith) [reply Doc/reference/expressions.rst]: Subtlety: It can't happen after the event loop exits, thanks to the async gen gc hooks.

**Comment by 1st1** [comment]
There's a bunch of good feedback from @belm0 -- please address and I think the PR is OK to be merged now. We can polish it in follow up PRs. @aeros please re-review.

File: Doc/library/contextlib.rst
Comment (achimnol) [reply Doc/library/contextlib.rst]: Fixed!

File: Lib/contextlib.py
Comment (achimnol) [reply Lib/contextlib.py:317]: Fixed!

File: Doc/reference/expressions.rst
Comment (achimnol) [reply Doc/reference/expressions.rst]: Updated!

File: Doc/library/contextlib.rst
Comment (achimnol) [reply Doc/library/contextlib.rst]: Replaced the example with #22640!

File: Misc/NEWS.d/next/Library/2020-07-19-20-10-41.bpo-41229.p8rJa2.rst
Comment (achimnol) [reply Misc/NEWS.d/next/Library/2020-07-19-20-10-41.bpo-41229.p8rJa2.rst]: Updated!

**Comment by aeros** [comment]
> please address and I think the PR is OK to be merged now. We can polish it in follow up PRs. @aeros please re-review. @aeros please re-review.

Sounds good, I'll look through it as soon as I get the chance. If I'm not able to by this weekend, I'll definitely be able to during the core sprint next week.

File: Lib/contextlib.py
Comment (belm0) [inline Lib/contextlib.py]: Is this description too specific?  `aclosing` can be used with any object having `aclose()`, not necessarily a generator.

File: Lib/contextlib.py
Comment (belm0) [inline Lib/contextlib.py]: Sorry, I said the `await` should be here, but that was incorrect.

The object passed to `aclosing()` is just taken as is, returned by `__aenter__`, and has `aclose()` called in `__aexit__`.

File: Lib/contextlib.py
Comment (achimnol) [reply Lib/contextlib.py]: Updated a little to embrace other cases.

File: Lib/contextlib.py
Comment (achimnol) [reply Lib/contextlib.py]: Oh, I also missed to check it. Reverted here.

**Review by belm0 — APPROVED** [review]
Thank you!

**Review by corona10 — COMMENTED** [review]
@aeros Gentle ping~

**Review by aeros — APPROVED** [review]
Thanks for the reminder @corona10, for the suggestions @belm0, and for the changes made @achimnol! LGTM, I'll proceed with merging.

(The only remaining component that would also be nice is a mention of the new `contextlib.aclosing()` in the 3.10 whatsnew document. But that can be done in a separate PR -- if you open one, I'd suggest linking to it here and requesting a review from me)

**Comment by uburuntu** [comment]
😒
https://github.com/python/cpython/pull/21285
https://bugs.python.org/issue41197

**Pull request merged**
